File: flamingo2-web/src/main/java/org/opencloudengine/flamingo2/web/system/MenuController.java
/**
 * Copyright (C) 2011 Flamingo Project (http://www.cloudine.io).
 * <p/>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * <p/>
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * <p/>
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.opencloudengine.flamingo2.web.system;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.commons.lang.exception.ExceptionUtils;
import org.opencloudengine.flamingo2.core.rest.Response;
import org.opencloudengine.flamingo2.core.security.SessionUtils;
import org.opencloudengine.flamingo2.web.configuration.ConfigurationHelper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.*;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

@Controller
@RequestMapping("/system/menu")
public class MenuController {

    /**
     * SLF4J Logging
     */
    private Logger logger = LoggerFactory.getLogger(MenuController.class);

    @Autowired
    private ObjectMapper objectMapper;

    @Autowired
    MenuService service;

    @RequestMapping(value = "select", method = RequestMethod.GET)
    @ResponseStatus(HttpStatus.OK)
    @ResponseBody
    public List<Map<String, Object>> select(@RequestParam Map<String, Object> params) {
        try {
            params.put("locale", ConfigurationHelper.getHelper().get("default.locale"));
            params.put("level", SessionUtils.getLevel());
            return service.select(params);
        } catch (Exception e) {
            logger.debug(e.getMessage());
            return null;
        }
    }

    @RequestMapping(value = "selectNode", method = RequestMethod.GET)
    @ResponseStatus(HttpStatus.OK)
    @ResponseBody
    public Response selectNode(@RequestParam Map<String, Object> params) {

        Response response = new Response();
        try {
            params.put("locale", ConfigurationHelper.getHelper().get("default.locale"));
            response.getList().addAll(service.selectNode(params));
            response.setTotal(response.getList().size());
            response.setSuccess(true);
        } catch (Exception ex) {
            response.setSuccess(false);
            response.getError().setMessage(ex.getMessage());
            if (ex.getCause() != null) response.getError().setCause(ex.getCause().getMessage());
            response.getError().setException(ExceptionUtils.getFullStackTrace(ex));
            logger.debug(ex.getMessage());
        }
        return response;
    }

    @RequestMapping(value = "save", method = RequestMethod.POST)
    @ResponseStatus(HttpStatus.OK)
    @ResponseBody
    public Response save(@RequestBody Map<String, Object> params) {
        Response response = new Response();
        try {

            ArrayList newRecords = objectMapper.readValue(params.get("newRecords").toString(), ArrayList.class);
            ArrayList updatedRecords = objectMapper.readValue(params.get("updatedRecords").toString(), ArrayList.class);
            ArrayList removedRecords = objectMapper.readValue(params.get("removedRecords").toString(), ArrayList.class);

            for (Object record : newRecords) {
                service.insert((Map<String, Object>) record);
            }

            for (Object record : updatedRecords) {
                service.update((Map<String, Object>) record);
            }

            for (Object record : removedRecords) {
                service.delete((Map<String, Object>) record);
            }

            response.setSuccess(true);
        } catch (Exception ex) {
            response.setSuccess(false);
            response.getError().setMessage(ex.getMessage());
            if (ex.getCause() != null) response.getError().setCause(ex.getCause().getMessage());
            response.getError().setException(ExceptionUtils.getFullStackTrace(ex));
            logger.debug(ex.getMessage());
        }
        return response;
    }
}

File: flamingo2-web/src/main/webapp/resources/app/view/system/menu/MenuController.js
/*
 * Copyright (C) 2011 Flamingo Project (http://www.cloudine.io).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
Ext.define('Flamingo2.view.system.menu.MenuController', {
    extend: 'Ext.app.ViewController',
    alias: 'controller.systemMenuViewController',

    onAfterrender: function () {
    },

    onBtnRefreshClick: function () {
        var menu = this.lookupReference('trpMenu');

        menu.getStore().load({
            callback: function () {
                menu.expandAll();
            }
        });
    },

    onBtnCollapseClick: function () {
        this.lookupReference('trpMenu').collapseAll();
    },

    onBtnExpandClick: function () {
        this.lookupReference('trpMenu').expandAll();
    },

    onBtnAddClick: function () {
        var me = this;
        var refs = this.getReferences();
        var selection = refs.trpMenu.getSelectionModel().getSelection();
        var id = selection[0].get('id');
        var depth = selection[0].get('depth');
        var cnt = refs.grdNode.getStore().getCount();

        if (depth == 2) {
            Ext.MessageBox.show({
                title: message.msg('common.check'),
                message: message.msg('menu.msg.can_make_2depth'),
                buttons: Ext.MessageBox.OK,
                icon: Ext.MessageBox.INFO
            });
            return;
        }

        var params = {
            prnts_menu_id: selection[0].get('id'),
            sort_ordr: cnt + 1,
            use_yn: 'Y'
        };

        refs.grdNode.getStore().insert(cnt, params);
    },

    onBtnSaveClick: function () {
        var me = this;
        var refs = this.getReferences();
        var store = refs.grdNode.getStore();

        if (!store.isValid()) {
            return;
        }

        var params = store.getAllJson();

        invokePostByMap(
            CONSTANTS.SYSTEM.MENU.SAVE,
            params,
            function (response) {

                var obj = Ext.decode(response.responseText);
                if (obj.success) {
                    Ext.MessageBox.show({
                        title: message.msg('common.success'),
                        message: message.msg('menu.msg.save_menu'),
                        buttons: Ext.MessageBox.OK,
                        icon: Ext.MessageBox.INFO
                    });

                    refs.grdNode.getStore().removeAll();
                    me.onBtnRefreshClick();
                }
                else {
                    Ext.MessageBox.show({
                        title: message.msg('common.error'),
                        message: message.msg('menu.msg.save_error') + '<br>' + obj.error.cause,
                        buttons: Ext.MessageBox.OK,
                        icon: Ext.MessageBox.INFO,
                        fn: function (e) {
                            return false;
                        }
                    });
                }
            },
            function (response) {
                Ext.MessageBox.show({
                    title: message.msg('common.warn'),
                    message: format(message.msg('common.msg.server_error'), config['system.admin.email']),
                    buttons: Ext.MessageBox.OK,
                    icon: Ext.MessageBox.WARNING
                });
            }
        );
    },

    onBtnRemoveClick: function () {
        var me = this;
        var refs = me.getReferences();
        var selection = refs.grdNode.getSelectionModel().getSelection();

        if (!selection.length) {
            Ext.MessageBox.show({
                title: message.msg('common.check'),
                message: message.msg('menu.msg.select_delete_row'),
                buttons: Ext.MessageBox.OK,
                icon: Ext.MessageBox.INFO
            });
            return;
        }

        refs.grdNode.getStore().remove(selection[0]);
    },

    onTrpMenuSelect: function (rowModel, record, index) {
        var me = this;
        var refs = me.getReferences();
        var menu_id = record.get('id');

        refs.grdNode.getStore().load({
            params: {
                prnts_menu_id: menu_id
            }
        });
    }
});